src/Timing.cc

```diff
@@ -0,0 +1,334 @@
+// All rights reserved.
+//
+// BSD 3-Clause License
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+//   list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include "ord/Timing.h"
+
+#include <tcl.h>
+
+#include "db_sta/dbNetwork.hh"
+#include "db_sta/dbSta.hh"
+#include "odb/db.h"
+#include "ord/OpenRoad.hh"
+#include "sta/Search.hh"
+//#include "ord/Tech.h"
+#include "ord/Design.h"
+#include "sta/Corner.hh"
+#include "sta/Liberty.hh"
+#include "sta/TimingArc.hh"
+#include "sta/TimingRole.hh"
+#include "utl/Logger.h"
+
+namespace ord {
+
+Timing::Timing(Design* design) : design_(design)
+{
+}
+
+sta::dbSta* Timing::getSta()
+{
+  auto app = OpenRoad::openRoad();
+  return app->getSta();
+}
+
+std::pair<odb::dbITerm*, odb::dbBTerm*> Timing::staToDBPin(const sta::Pin* pin)
+{
+  ord::OpenRoad* openroad = ord::OpenRoad::openRoad();
+  sta::dbNetwork* db_network = openroad->getDbNetwork();
+  odb::dbITerm* iterm;
+  odb::dbBTerm* bterm;
+  db_network->staToDb(pin, iterm, bterm);
+  return std::make_pair(iterm, bterm);
+}
+
+bool Timing::isEndpoint(odb::dbITerm* db_pin)
+{
+  sta::Pin* sta_pin = getSta()->getDbNetwork()->dbToSta(db_pin);
+  return isEndpoint_(sta_pin);
+}
+
+bool Timing::isEndpoint(odb::dbBTerm* db_pin)
+{
+  sta::Pin* sta_pin = getSta()->getDbNetwork()->dbToSta(db_pin);
+  return isEndpoint_(sta_pin);
+}
+
+bool Timing::isEndpoint_(sta::Pin* sta_pin)
+{
+  auto search = sta::Sta::sta()->search();
```

Comment:
Use getSta() instead of accessing this singleton (throughout).  dbSta inherits from  Sta.